blogs/posts/2024-08-08-map-and-nest/map-and-nest.qmd

```diff
@@ -0,0 +1,127 @@
+- `purrr::map()`
+- `purrr::walk()`
+
+
+## Example on NHS workforce statistics 
+
+```{r}
+#| echo: false
+source("helper-functions.R")
+```
+
+The [NHS workforce statistics](https://digital.nhs.uk/data-and-information/publications/statistical/nhs-workforce-statistics) are official statistics published monthly for England.
+
+```{r}
+staff_group <- readRDS(file = "workforce_staff_group.rds")
+```
+
+I want to perform an analysis for each of the 42 integrated care systems (ICS). The {tidyr} `nest()` function creates a list-column, where each cell contains a mini dataframe for each grouping.
+
+Let's group by ICS, and call the nested data column `raw_data`.
+```{r}
+group_by_ics <- staff_group |>
+    tidyr::nest(raw_data = -ics_name)
+```
+
+The new column is a list-column, with each cell containing an entire tibble of data relating to that individual ICS.
+
+```{r}
+#' echo: false
+head(group_by_ics)
+```
+
+We can grab these mini datasets in the usual way and explore them interactively.
+
+```{r}
+group_by_ics$raw_data[[1]]
+```
+
+Next, let's apply some simple processing, say converting absolute numbers into percentages, to each of the ICSs in turn. We use `mutate()` to create a new list-column `staff_percent` and `map()` to apply the processing function to each cell in turn. ^[In this example, we actually didn't need to nest first. We could have performed the `mutate()` step on the full dataset.]
+
+```{r}
+processed_staff <-
+group_by_ics |>
+    dplyr::mutate(
+        staff_percent = purrr::map(raw_data, convert_percent)
+    )
+```
+
+Where I think this map-and-nest process really comes into it's own us creating plots. Often, I find myself wanting to create a couple of different plots for each grouping, and then optionally save the plots with sensible names. Particulary in the analysis stage, I like having these plots in the same row as the raw data, so I can quickly compare and validate.
+
+I've created [two functions](https://github.com/The-Strategy-Unit/data_science/tree/main/blogs/posts/2024-08-08-map-and-nest/helper-functions.R) `plot_barchart()` and `plot_waffle()`, which take the data and create charts. Again, using `mutate()` I can create a new column called `barchart` and I can `map()` the function `plot_barchart()`, applying it to each row at a time.
+
+```{r}
+graphs <-
+processed_staff |>
+    dplyr::mutate(
+        barchart =  purrr::map(staff_percent, plot_barchart)
+    ) 
+```
+
+The resulting column `barchart` is again a list-column, but this time instead of containing a tibble, it holds a ggplot object. A whole ggplot in a single cell. ^[This totally blew my mind the first time I saw it 🤯.]
+
+If we want to pass two arguments to our function, we can replace `map()` with `map2()`. Here we're using `map2()` to pass the `ics_name` column to use as a title in our waffle plot.
+
+```{r}
+graphs <-
+processed_staff |>
+    dplyr::mutate(
+        waffle =  purrr::map2(raw_data, ics_name, ~plot_waffle(.x, title = .y))
```

Comment:
I wonder if adding the arg names might make clearer the connection from `raw_data` and `ics_name` to `.x` and `.y` for folks who haven't done a `map2()` before?

```suggestion
        waffle =  purrr::map2(
            .x = raw_data, 
            .y = ics_name, 
            ~plot_waffle(.x, title = .y)
        )
```